<big>ПМ24-5 Белов Фёдор </big>

In [2]:
import numpy as np
def rrstr(x,n): # округление до n знаков после запятой
    fmt = '{:.'+str(n)+'f}'
    return fmt.format(x).replace('.',',')

1. Для каждого $k=1,\dots,n$ обозначим $e_k$ вектор-столбец размерности $n$, такой, что 
его элемент с номером $k$ равен 1, а все остальные элементы равны 0. 
Пусть $\sigma: \{1,\dots,n\}\rightarrow\{1,\dots,n\}$  – некоторая перестановка первых $n$ натуральных чисел.
Матрица $A$ однозначно определяется условиями: $Ae_k=e_{\sigma(k)}, k=1,\dots,n$. 
Известно, что $n=12$ и результатом применения $\sigma$ к последовательности $1,\dots,12$ будет следующий ряд чисел:
$1$, $7$, $6$, $10$, $12$, $3$, $11$, $5$, $9$, $4$, $8$, $2$. $\:\:$
Найдите:
1) след матрицы $A^{5}$;
2) наименьшее натуральное $m$, такое, что след матрицы $A^m$ равен 12.<br/>

In [3]:
s1 = [1, 7, 6, 10, 12, 3, 11, 5, 9, 4, 8, 2]
s0 = [x-1 for x in s1]
s1, s0

([1, 7, 6, 10, 12, 3, 11, 5, 9, 4, 8, 2],
 [0, 6, 5, 9, 11, 2, 10, 4, 8, 3, 7, 1])

In [18]:
from sympy.combinatorics.permutations import Permutation
p = Permutation(s0)
p5 = p**5
c = 0

for i in range(12):
    if i == p5(i):
        c += 1
c

2

In [8]:
p.order()

6

2. Пусть $\lambda$ – комплексный корень из 1 степени 37, такой, что его мнимая часть больше 0, 
а действительная часть максимальна среди подобных корней с положительной мнимой частью.
Даны матрицы: $I = \pmatrix{1 & 0 \\ 0 & 1}$,
$A = \pmatrix{4 & 6 & \\ 4 & 7}\!.\;$
Найдите определитель матрицы $A + \lambda I$ и укажите в ответе:
1) действительную часть этого определитля;
2) его мнимую часть.<br/>

In [29]:
import math
from sympy import *
import numpy as np
tetha = 2 * math.pi / 37
lam = np.exp(1j * tetha)
I = eye(2)
A = Matrix([[4, 6],
           [4, 7]])
rrstr(re((A + lam * I).det()), 3), rrstr(im((A + lam * I).det()), 3)

('15,785', '2,192')

Ответ: 15,785; 2,192

3. Для многочлена $p(x)=x^{14}-x^{12}+x^{11}+x^{10}-x^{9}+x^{7}+2x^{4}+2$ найдите все его $14$ корней: $z_1, \dots, z_{14} \in \Bbb C$.
В ответе укажите:
1) сумму модулей $|z_1|+ \ldots +|z_{14}|$;
2) действительную часть корня, у которого эта действительная часть максимальна.<br/>

In [33]:
coef = [1, 0, -1, 1, 1, -1, 0, 1, 0, 0, 2, 0, 0, 0, 2]
roots = np.roots(coef)
rrstr(np.sum(abs(roots)), 3), rrstr(max(roots.real), 3)

('14,763', '1,062')

4. Дана матрица: 
$A = \pmatrix{18 & 27 & 19\\ 20 & 15 & 24 \\ 28 & 29 & 25}$.
Пусть $Z$ – собственное значение матрицы $A$ с наибольшей мнимой частью.
Найдите $Z$ и соответствующий этому собственному значению собственный вектор $X$, 
первая координата которого равна 1, $X=(1,U,V)$.
В ответе укажите:
1) мнимую часть $Z$;
2) действительную часть $U$.<br/>

In [39]:
A = Matrix([[18, 27, 19],
           [20, 15, 24],
           [28, 29, 25]])
eigenvects = A.eigenvects()
for val, alg, vects in eigenvects:
    val = val.evalf()
    vects = [x.evalf() for x in vects][0]
    vects = [(x/vects[0]).evalf() for x in vects]
    display(val)
    display(vects)

-5.13364248554728 - 1.25115300092831*I

[1.00000000000000,
 -0.644712952763244 - 0.170849065560288*I,
 -0.301389092681036 + 0.176935356273657*I]

-5.13364248554728 + 1.25115300092831*I

[1.00000000000000,
 -0.644712952763244 + 0.170849065560288*I,
 -0.301389092681036 - 0.176935356273657*I]

68.2672849710946

[1.00000000000000, 0.955626448423584, 1.28765109808725]

5. На двумерной координатной плоскости заданы точки: 
$A(37,19)$, $B(19,43)$ и $C(3,4)$.
Пусть $D(X,Y)$ – ближайшая к $A$ точка прямой $BC$.
Найдите точку $D$ и расстояние $R$ от $A$ до $D$.
Укажите в ответе:
1) X, первую координату точки D;
2) Y, вторую координату точки D;
3) R, расстояние от A до D.<br/>

In [43]:
A = np.array([37, 19])
B = np.array([19, 43])
C = np.array([3, 4])
BA = A - B
BC = C - B
BABC = np.dot(BA, BC)
BCBC = np.dot(BC, BC)
t = BABC/BCBC
D = B + t*BC
D

array([13.16544738, 28.778278  ])

In [45]:
R = ((A[0] - D[0])**2 + (A[1] - D[1])**2)**0.5
R

25.762387680896772